---
title: "Introduction to Scientific Computing for Biologists"
subtitle: |
  | ISCB20.09 - R for Bioinformatics
  | An Introduction to `seqinr` and `Biconductor`
author: | 
  | Md. Jubayer Hossain
  | https://jhossain.me/
  | jubayer@hdrobd.org 
institute: | 
  | Founder 
  | Health Data Research Organization
  | Lead Organizer
  | Scientific Computing for Biologists 
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  beamer_presentation:
    highlight: espresso
    fonttheme: 'structurebold'
classoption: 'aspectratio=169'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, comment=NA)
```

# Section-1: Introduction to "seqinr"

## `seqinr`: Package Introduction{.t}
- Exploratory data analysis and data visualization for biological sequence (DNA and protein) data.

```r
install.packages("seqinr")
```
## Reading and Exploring Data{.t}
- Reading `FASTA`, `FASTQ` Files. 
- Exploring DNA and Protein Sequence.

## DNA Sequence Statistics{.t}
- Length of a DNA Sequence. 
- Nucleotide Frequency/Base Composition 
- Nucleotide Percentage 
- K-mer Analysis/DNA Words 
- Nucleotide Frequency Distribution Plot 
- GC Content of DNA 
- Local Variation in GC Content / Sliding Window Analysis
- Dot Plot

# Section-2: Introduction to Bioconductor
## What is Bioconductor?{.t}
- Bioconductor is a free, open source and open development software project for the analysis and comprehension of genomic data generated by wet lab experiments in molecular biology.
- Bioconductor is based primarily on the statistical R programming language, but does contain contributions in other programming languages.
- Website: https://www.bioconductor.org/

## Installing Bioconductor{.t}
Biconductor has its own repository, way to install packages, and each release is designed
to work with a speci c version of R.
```r
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(version = "3.12")
```

## What do we measure and why?{.t}
- Structure: elements, regions, size, order, relationships
- Function: expression, levels, regulation, phenotypes

## Useful Online Resources{.t}	
- https://kasperdanielhansen.github.io/genbioconductor/
- https://kasperdanielhansen.github.io/genbioconductor/html/Online_Resources.html
- https://www.datacamp.com/courses/introduction-to-bioconductor-in-r

## Introduction to `Biostrings` {.t}
- Memory efficient to store and manipulate sequence of characters. 
- Containers that can be inherited. 
- The BString class comes from big string
```r
showClass("XString")
showClass("BString")
showClass("BStringSet")
```
## Load Biostrings Package 
```{r}
# Load Biostrings
library(Biostrings)
```

## Biostring Alphabets{.t}
```{r}
DNA_BASES # DNA 4 Bases 
RNA_BASES # RNA 4 Bases 
AA_STANDARD # 20 Amino Acids(AA)
```

## Biostring Alphabets(Cont..){.t}
```{r}
DNA_ALPHABET # Contains IUPAC_CODE_MAP
RNA_ALPHABET # Contains IUPAC_CODE_MAP
AA_ALPHABET  # Contains AMINO_ACID_CODE
```

## Transcription: DNA to RNA{.t}
```{r}
# DNA Single String
dna_seq <- DNAString("ATGATCTCGTAATCCG")
dna_seq
```

```{r}
# Transcription DNA to RNA string
rna_seq <- RNAString(dna_seq)
rna_seq
```

## Translation: RNA to Amino Acids(AA){.t}
```{r}
# Translation RNA to Amino Acids(AA)
aa_seq <- translate(rna_seq)
aa_seq
```

## Shortcut Translation: DNA to Amino Acids(AA){.t}
```{r}
dna_seq <- DNAString("ATGATCTCGTAATCCG")
# Translate
translate(dna_seq)
```

## Single  vs Set{.t}
- XString to store a single sequence
  - BString for any string
  - DNAString for DNA
  - RNAString for RNA
  - AAString for amino acids
- XStringSet for many sequences
  - BStringSet
  - DNAStringSet
  - RNAStringSet
  - AAStringSet

## Creating and Collecting `stringSet` {.t}
```{r}
covid19 <- readDNAStringSet("covid19.fasta")
length(covid19) # The set contains only one sequence
width(covid19) # Bases 
```

## Creating and Collecting `stringSet`(Cont..){.t}
```{r}
# To collate the sequence use unlist
covid19_seq <- unlist(covid19)
length(covid19_seq)

# width(covid19_seq)
# Error unable to find width for "DNAString"
```

## From a Single Sequence to a Set{.t}
```{r}
# To create a new set from a single sequence
single_seq <- DNAStringSet(covid19_seq, 
                           start = c(1, 101, 201), 
                           end = c(100, 200, 300))
covid19_seq
```
```{r}
length(single_seq)
width(single_seq)
```

## Complement Sequence{.t}
```{r}
# Create a DNA Sequence 
dna_seq <- DNAString("ATGATCTCGTAATTCCGGAA")
dna_seq
```
```{r}
# Complement
complement(dna_seq)
```


## Rev a Sequence{.t}
```{r}
# Create a DNA Sequence 
dna_seq <- DNAString("ATGATCTCGTAATTCCGGAA")
dna_seq
```
```{r}
# Rev
rev(dna_seq)
```


## Reverse a Sequence{.t}
```{r}
dna_seq
# Reverse
reverse(dna_seq)
```


## Reverse Complement{.t}
```{r}
# Reverse Complement 
reverseComplement(dna_seq)
# Using two functions together
reverse(complement(dna_seq))
```

## Sequence Alignment{.t}
- Pairwise global alignment of DNA sequences using the Needleman-Wunsch algorithm

## Loading and Exploring Genome{.t}
```r
# Length
length()
# Names 
names() 
# Sequence Info 
seqinfo()
# Sequence Level Info 
seqlevels()
# Sequence Length 
seqlengths()
# Sequence Names 
seqnames()
# Sequence Style 
seqlevelsStyle()
```
